### PR TITLE
fix topic special tags

### PR DESCRIPTION
**Pull request by lunny**
This will filter special tags on topics.

### CODE REVIEW

**Comment by jonasfranz** [comment]
If I add the topic `"><iframe><div class="` it will result in:
screenshot-2018-5-24 xss

If I try to remove the same topic which I added before your PR will result in not removing it.
In the console this error appears:
`Error: Syntax error, unrecognized expression: [data-value=""><iframe><div class=""] jquery.min.js:2:12731`

**Comment by axifive** [comment]
I think for topics we need to add restrictions on the used symbols like _GitHub_ topics and _StackOverflow_ [tags](https://stackoverflow.com/help/privileges/create-tags).

**Rules:**
1. All topics are in lowercase
2. Can be used only `[a-z0-9_+.#-]` symbols
3. Length limitation - `{1, 35}` chars
4. Number of topics added to the repository - max 25 topics
... maybe something else

**Review by techknowlogick — APPROVED** [review]
I'm approving this because it gets rid of the security issue. A different PR can be created later to limit what can be in a tag.

File: models/topic.go [hunk 1]
```diff
@@ -113,7 +118,8 @@ func SaveTopics(repoID int64, topicNames ...string) error {
 
 	var addedTopicNames []string
 	for _, topicName := range topicNames {
-		if strings.TrimSpace(topicName) == "" {
+		topicName = validTopic(topicName)
+		if topicName == "" {
```
Comment (bkcsoft) [inline models/topic.go:122]:
```go
if topicName = validTopic(topicName); len(topicName) == 0 {
```

File: models/topic.go [hunk 2]
```diff
@@ -133,6 +139,11 @@ func SaveTopics(repoID int64, topicNames ...string) error {
 	for _, t := range topics {
 		var found bool
 		for _, topicName := range topicNames {
+			topicName = validTopic(topicName)
+			if topicName == "" {
```
Comment (bkcsoft) [inline models/topic.go:143]: same as above

File: public/js/index.js [hunk 3]
```diff
@@ -2301,12 +2301,23 @@ function initNavbarContentToggle() {
     });
 }
 
+function pasteFilter(e) {
```
Comment (bkcsoft) [inline public/js/index.js:2304]: What does this magical function do? Mind adding that to a comment above the function? :)

**Review by bkcsoft — CHANGES_REQUESTED** [review]
Only a few nit-picks

**Comment by jonasfranz** [comment]
@techknowlogick I cannot follow you because the security issue is not resolved as I stated in a previous comment. If  you think that this will resolve the problem, let me know it! 

Edit: It actually resolves the security problem but introduces a bug as I mentioned in the previous comment.

**Comment by techknowlogick** [comment]
@JonasFranzDEV yes. I give LG-TM because bug happens when a user has the `"`'s in their tag (at least that's the only way I can trigger it in my build of this branch), which hopefully is an uncommon character for a tag. This is why I am ok with the PR as it resolves the security, and still improves upon existing behavior.

~It is almost my lunch hour, and so I might be able to send a patch to Lunny.~ Sadly I wasn't able to accomplish this.

File: public/js/index.js (hunk 3 above)
Comment (techknowlogick) [reply public/js/index.js:2304]: It replaces all tags (substrings on the form <...>) with a space, " " when a user pastes, but I agree there should be a comment.

**Comment by techknowlogick** [comment]
@JonasFranzDEV what if in the `validTopic` function, the list of conditions that @axifive made were added?

**Comment by axifive** [comment]
@techknowlogick, I'm working on the full solution with front/back validation. And I also found repo topics duplication bug. I'll try to finish this on the weekend.

**Comment by lunny** [comment]
closed by #4258

**Pull request closed without merging**